design_notes.md

```diff
@@ -0,0 +1,227 @@
+  differ from the dimensions of the coordinates used to build the index).
+
+- It is not possible to mix orthogonal and point-wise indexing for indexers
+  given in ``ds.xoak.sel()``.
+
+- ``method='nearest'`` is implicitly assumed here.
+
+### Xarray custom indexes
+
+One important point of Xarray's development roadmap is the addition of [flexible
+indexes](http://xarray.pydata.org/en/stable/roadmap.html#flexible-indexes). Once
+this feature is available in xarray, the accessors mentioned here above should
+be depreciated in favor of custom index classes that can be used directly in
+xarray. Having an API close to xarray's current index creation and data selection
+methods will help towards a smooth transition.
+
+### Utility functions
+
+For convenience, xoak may provide some utility functions, e.g., to transform
+lat/lon coordinates into XYZ cartesian coordinates.
+
+### Low-level API
+
+In case xoak will eventually have its own index implementations, it should also
+provide a low-level API for reusing the indexes outside of Xarray.
+Alternatively, those implementations could be maintained in a 3rd-party package.
+
+### Indexing operations not supported by xarray
+
+Data partitioning trees often allow queries other than just range and/or nearest
+neighbor searches, e.g., k-nearest neighbor search, radius-based search, point
+density/count, compute distances, selection of objects, etc. How to expose those
+queries in xoak remains an open question.
+
+### Other operations supported by xarray
+
+Data partitioning trees have also the potential to be reused in important xarray
+features other than data selection such as, e.g., group-by operations. How to
+handle those operations with xoak remains an open question too.
+
+## 5. Internals
+
+Instead of directly reusing tree-based indexes in 3rd party packages, we might
+need to write some thin wrappers around them in order to:
+
+- store coordinate dimensions
+- properly handle support for orthogonal vs. point-wise indexing
+- re-arrange the data (coordinate or indexer values) without copy
+- index serialization (see below)
+- etc.
+
+## 6. Index serialization / reconstruction
+
+Building a tree index can be time consuming. xoak should allow saving/loading
+indexes. This could be done independently of the data in xarray objects by
+pickling the index objects.
+
+A more general solution has still to be found on the xarray side. This will
+probably be addressed during the implementation of flexible indexes.
+
+## 7. dask support -- distributed / out-of-core indexes
+
+Most data partitioning trees are centralized structures, making it hard to use
+in the context of distributed or out-of-core computation.
+
+There are two possible cases:
+
+- large coordinate data
+- large indexer data
+
+Possible issues are:
+
+- loading all coordinate data into memory
+- tree construction time
+- tree storage in memory
+- loading all indexer data into memory
+- selection (query) time
+
```

Comment:
We should mention definite use cases (LLC4320, ARGO, Surface Drifters) from the climate and ocean sciences, derive data volumes for coordinates, indexers, and trees, and give a statement on which issues are the most likely to limit performance in real-world use cases.

Maybe we put this in a section `## 8. Resource estimates for climate-related use cases`?

Reply:
Yes, that would be welcome. I'd simply suggest `## 8. Use cases`.

Reply:
Let's do this in a separate PR, then.